### PR TITLE
feat: add graphql

### PR DESCRIPTION
**Pull request by jonowu**
## Purpose

Refactor the app to use graphql.

## Changes

You will need to run `npm install` when testing this branch.
You can test graphql queries here: http://sa4si-admin.herokuapp.com/graphql

## Checklist

- [ ] Tested on Android
- [ ] Tested on iOS

### CODE REVIEW

File: utils/apolloClient.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import { api } from '../data/api';
+
+const cache = new InMemoryCache();
+
+const httpLink = new HttpLink({
+  uri: `${api}/graphql`,
+});
+
+const authLink = setContext(async (_, { headers }) => {
+  const token = await AsyncStorage.getItem('token');
```
Comment (jonowu) [inline utils/apolloClient.js:14]: Pulls the bearer token from async storage

File: screens/sdgs/index.js [hunk 2]
```diff
@@ -12,22 +14,46 @@ const TileContainer = styled.TouchableOpacity`
...
 
   const Item = ({ sdg }) => (
     <TileContainer onPress={() => navigation.navigate('SDG', { sdg: sdg })}>
       <Image source={sdg.src} resizeMode="contain" style={{ height: '100%', width: '100%' }} />
     </TileContainer>
   );
 
+  const mergedSdgs = _.map(sdgs, (item) => {
+    return _.assign(item, _.find(sdgData, ['id', item.number]));
+  });
```
Comment (jonowu) [inline screens/sdgs/index.js:40]: This combines the local array of sdg objects with the data from strapi (where they have the same sdg number/id).

File: screens/action/index.js [hunk 3]
```diff
@@ -15,35 +14,27 @@ function ActionScreen({ route, navigation }) {
...
-  const token = authContext.user.token;
 
-  const shareActionMessage = `#SA4SI - I made a difference by completing the action "${title}"!
-   Download the #SA4SI app to join me!`;
+  const shareActionMessage = `I made a difference by completing the action "${title}"!
+    Download the #SA4SI app to join me!`;
+
+  const CREATE_ENTRY = gql`
+    mutation {
+      createEntry(input: { data: { action: ${id}, user: ${userId} } }) {
```
Comment (jonowu) [inline screens/action/index.js:23]: A mutation is like a post request. Here, we create a new entry that includes the action and user as data.

File: screens/actions/index.js [hunk 4]
```diff
@@ -1,74 +1,73 @@
...
-  const [completedActions, setCompletedActions] = useState([]);
-
   const authContext = useContext(AuthenticatedContext);
   const id = authContext.user.data.id;
-  const token = authContext.user.token;
+
+  const [completedActions, setCompletedActions] = useState([]);
+
+  const GET_USER_ACTIONS = gql`
+    query GetUserActions {
```
Comment (jonowu) [inline screens/actions/index.js:16]: This query gets the all the actions, and entries that the currently logged in user has submitted. This an advantage of graphql; we can query multiple datasets in a single request.

File: screens/news/index.js [hunk 5]
```diff
@@ -61,56 +58,55 @@ const TileTextDescription = styled.Text`
...
-        setLoading(false);
-      })
-      .catch((error) => {
-        alert(error.response.data.error);
-        console.log('An error occurred:', error.response);
-      });
-  }, []);
+  const GET_ARTICLES = gql`
+    query GetArticles {
+      articles {
```
Comment (jonowu) [inline screens/news/index.js:63]: This query simply gets all the articles, with only the relevant data that we need. This is a big advantage of graphql; we only get back the data we ask for.

**Review by BenjaminYong — CHANGES_REQUESTED** [review]
Looks good, works well on my phone.
Can only suggest a screen or styling to center the loading indicators 👍

**Pull request merged**